File: solutions/_47_permutations_ii/solution.py
from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def backtrack(perm):
            if len(perm) == len(nums):
                res.append(perm[:])
                return
            for i in range(len(nums)):
                if visited[i] or (
                    i > 0 and nums[i] == nums[i - 1] and not visited[i - 1]
                ):
                    continue
                visited[i] = True
                perm.append(nums[i])
                backtrack(perm)
                visited[i] = False
                perm.pop()

        res = []
        visited = [False] * len(nums)
        nums.sort()
        backtrack([])
        return res


# https://leetcode.com/problems/permutations-ii/

File: solutions/_47_permutations_ii/solution.cpp

#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> nums;
    vector<vector<int>> res;
    vector<bool> visited;

    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());
        this->nums = nums;
        this->visited = vector<bool>(nums.size(), false);
        vector<int> perm;
        backtrack(perm);
        return res;
    }

    void backtrack(vector<int> &perm)
    {
        if (perm.size() == nums.size())
        {
            res.push_back(perm);
        }
        else
        {
            for (int i = 0; i < nums.size(); i++)
            {
                if (visited[i])
                    continue;
                if (i > 0 && nums[i] == nums[i - 1] && !visited[i - 1])
                    continue;
                visited[i] = true;
                perm.push_back(nums[i]);
                backtrack(perm);
                visited[i] = false;
                perm.pop_back();
            }
        }
    };
};

/*
https://leetcode.com/problems/permutations-ii/
*/